### PR TITLE
TLS Phase 1: Poly1305 one-time authenticator (RFC 8439) + KATs

### PR DESCRIPTION
**Pull request by kamrankhan78694**
## Subsystem
`tls` crypto — second hand-written primitive for `TLS_CHACHA20_POLY1305_SHA256` (follows the ChaCha20 PR). Pairs with ChaCha20 to build the AEAD next.

## What
- **`src/tls/poly1305.{c,h}`** — `poly1305_mac(key[32], msg, len, tag[16])`, the RFC 8439 §2.5 one-time authenticator.
  - Field arithmetic **mod 2^130-5 in five 26-bit limbs** with 64-bit products — the standard constant-time formulation: no secret-dependent branch, index, or table lookup.
  - Correct handling of the final **partial block** (explicit `0x01` high bit placed after the data) and the **constant-time `h >= p` final reduction** (compute `g = h − p`, select via a mask, no branch).
  - `r` clamp `0x0ffffffc0ffffffc0ffffffc0fffffff`; tag = `(h + s) mod 2^128`.

## Safety / scope
Purely **additive**, gated behind `WEBLIB_TLS` (in `WEBLIB_SOURCES_TLS`). Default build compiles no Poly1305 — `nm build/libweblib.a` shows no `poly1305` symbol; default `ctest` unchanged. No existing code touched.

## Tests (6 KATs)
- **RFC 8439 §2.5.2** — the 34-byte message (two full blocks **+ a 2-byte partial**, so both paths are covered).
- **empty** message → tag == `s`.
- **16-byte exact** block (no partial tail).
- **48-byte** three-block message.
- **RFC A.3 #1** — all-zero key + 64-byte zero message → all-zero tag.
- **r == 0** (clamped) → accumulator stays 0 → tag == `s`.

RFC-published vectors plus edge cases were **cross-checked against an independent big-integer reference** before hardcoding. **Negative control:** a wrong `r` clamp fails exactly the nonzero-`r` vectors (§2.5.2/16B/48B) while the r=0/empty/all-zero cases correctly still pass — so the KATs pin the arithmetic.

## Validation
- TLS build (`-DWEBLIB_ENABLE_TLS=ON`): `ctest` **8/8**, warning-free; Poly1305 KATs pass under **ASan/UBSan** (no overflow/UB in the limb arithmetic).
- Default build (TLS OFF): **6/6** on the normal and ASan/UBSan trees — unchanged.

## Note
`poly1305_mac` wipes the stack copy of the last message block; the `r`/`h` values live in scalar locals that portable C can't reliably clear (register copies), so they're not wiped with an ineffective self-assignment — documented in the source.

## Next
The ChaCha20-Poly1305 AEAD (RFC 8439 §2.8) combining these two, with its §2.8.2 KAT — including the Poly1305 key derivation (ChaCha20 block 0) and the AAD/ciphertext/length MAC construction.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by kamrankhan78694 — COMMENTED** [review]
## Independent review

Reviewed against RFC 8439 §2.5, constant-time discipline, the reduction correctness, and scope isolation. **No blocking issues.**

### Verified
- **Correctness is pinned to the RFC and to edge cases the main vector misses.** §2.5.2 covers full + partial blocks; the extra cases specifically exercise the paths a single vector wouldn't: empty (h==0), an exact 16-byte boundary (no partial tail), a 3-block message, all-zero, and `r==0` (accumulator stays 0, so the final reduction takes its no-subtract branch). All six pass, and they were cross-checked against an independent big-integer reference before hardcoding.
- **The reduction is constant-time and correct.** `g = h + 5 − 2^130` with the `mask = (g4 >> 31) − 1` select is the standard branch-free "subtract p if h ≥ p"; the KATs (including values that do and don't need the subtraction) confirm it. Limb carries fold the 2^130 overflow via the `5*r` terms.
- **Partial-block handling matches the spec:** full blocks add 2^128 (`hibit = 1<<24`); the final short block places an explicit `0x01` at `buf[len]` and adds no extra high bit. `len == 0` is handled (loop body never runs → tag = s).
- **Memory / UB clean.** 64-bit products of 26-bit×26-bit limbs can't overflow; shifts are in range. ASan/UBSan pass. Gated behind `WEBLIB_TLS`; `nm` shows no `poly1305` in the default lib; default `ctest` unchanged.

### Note (documented, no change requested)
`r`/`h` are scalar locals; portable C can't reliably wipe register-resident values, so — unlike ChaCha20's `state`/`x` **arrays** — they aren't zeroed (an ineffective self-assign would be misleading). The stack message-block buffer *is* wiped. If we later want the accumulator/key material wiped too, the clean route is to hold them in a small array and `secure_zero` it — but I'd only do that carefully, since a wiping-motivated refactor of verified limb arithmetic carries more risk than the un-wiped scalars.

_🤖 Reviewed by [Claude Code](https://claude.com/claude-code)_

File: tests/test_tls_crypto.c [hunk 1]
```diff
@@ -11,10 +11,25 @@
...
+    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
+    return 0;
+}
+
+static void from_hex(const char *hex, uint8_t *out, size_t out_len) {
+    size_t i;
+    for (i = 0; i < out_len; i++) {
+        out[i] = (uint8_t)((hexval(hex[i * 2]) << 4) | hexval(hex[i * 2 + 1]));
+    }
+}
```
[the remark this answers is not in the document]
Comment (kamrankhan78694) [reply tests/test_tls_crypto.c:42]: Fixed in 447caf6. `from_hex()` now checks for the NUL terminator before reading each nibble (never reads past the string end) and, on a short or non-hex input, prints a failure, marks `g_failures`, and zeroes the output; `hexval()` returns -1 for non-hex rather than silently mapping it to 0.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: tests/test_tls_crypto.c [hunk 2]
```diff
@@ -102,6 +117,48 @@ static void test_chacha20_encrypt(void) {
...
+static void test_poly1305(void) {
+    uint8_t key[32];
+    uint8_t tag[16];
+
+    /* §2.5.2: 34-byte message = two full blocks + a 2-byte partial. */
+    from_hex("85d6be7857556d337f4452fe42d506a80103808afb0db2fd4abff6af4149f51b", key, 32);
+    poly1305_mac(key, (const uint8_t *)"Cryptographic Forum Research Group", 34, tag);
+    check_hex("poly1305 (RFC 8439 2.5.2)", tag, 16, "a8061dc1305136c6c22b8baf0c0127a9");
+
+    /* Empty message: h stays 0, so tag == s (low 16 key bytes). */
```
[the remark this answers is not in the document]
Comment (kamrankhan78694) [reply tests/test_tls_crypto.c]: Correct — fixed the comment in 447caf6. The empty-message tag equals `s`, the HIGH 16 key bytes (16..31) = `0103...f51b`, not the low half (that's `r`).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: tests/test_tls_crypto.c [hunk 3]
```diff
@@ -102,6 +117,48 @@ static void test_chacha20_encrypt(void) {
...
+/* Poly1305 one-time authenticator — RFC 8439 §2.5.2 plus edge cases (empty,
+ * exact block boundary, multi-block, all-zero, and r==0 -> tag==s). */
+static void test_poly1305(void) {
+    uint8_t key[32];
+    uint8_t tag[16];
+
+    /* §2.5.2: 34-byte message = two full blocks + a 2-byte partial. */
+    from_hex("85d6be7857556d337f4452fe42d506a80103808afb0db2fd4abff6af4149f51b", key, 32);
+    poly1305_mac(key, (const uint8_t *)"Cryptographic Forum Research Group", 34, tag);
+    check_hex("poly1305 (RFC 8439 2.5.2)", tag, 16, "a8061dc1305136c6c22b8baf0c0127a9");
```
[the remark this answers is not in the document]
Comment (kamrankhan78694) [reply tests/test_tls_crypto.c]: Fixed in 447caf6. Both string-literal tests (§2.5.2 and r=0) now bind the literal to a `const char *msg` and pass `strlen(msg)` — no more hard-coded 34/36, so editing the literal can't desync the length.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
